### PR TITLE
3 new configurations added: /bun, /jsdoc, and /jsdoc-typescript

### PR DESCRIPTION
**Pull request by mangs**
**Pull Request Checklist**

- [x] I have read the [CONTRIBUTING](/CONTRIBUTING.md) document
- [x] Readme and changelog updates were made reflecting this PR's changes
- [x] Increase the project version number in `package.json` following [Semantic Versioning](http://semver.org/)

**Changes Included**

- Version bump to `2.0.0`
- **Breaking Changes**
  - Node.js `18.18.0` is now the minimum supported runtime version because of dependency requirements
  - ESLint version `8.56.0` is now the minimum supported version
    - Version `9.x` cannot be supported yet due to a dependency on the `eslint-config-airbnb*` packages and its strict peer dependency requirement of no higher than ESLint `8.x`
- 3 new configurations added
  - `/bun` for Bun support
  - `/jsdoc` for JSDoc support in JavaScript projects
  - `/jsdoc-typescript` for JSDoc support in TypeScript projects
- [Bun](https://bun.sh/) replaces NPM as a script runner and package manager for maintenance of this package
- When grouping module imports, [subpath imports](https://nodejs.org/api/packages.html#subpath-imports) are now treated as part of the 2nd or `Internal Imports` module imports grouping
- Upgrade dependencies to latest versions

### CODE REVIEW

File: lib/eslintPlaywrightConfig.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "plugins": ["eslint-plugin-playwright"],
-  "extends": ["plugin:playwright/recommended", "./eslintNodeConfig.json", "eslint-config-prettier"],
+  "extends": ["plugin:playwright/recommended"],
```
Comment (mangs) [inline lib/eslintPlaywrightConfig.json:3]: Node.js can no longer be assumed. Could be Bun or Node.

Comment (mangs) [reply lib/eslintPlaywrightConfig.json:3]: `eslint-config-prettier` only belongs as the last `extends` entry, so it should only ever be in one of the environment configs:
- Node
- Bun
- Browser

File: lib/eslintTypescriptConfig.json [hunk 2]
```diff
@@ -3,14 +3,13 @@
   "parserOptions": {
     "project": ["./tsconfig.json"]
   },
-  "plugins": ["@typescript-eslint/eslint-plugin", "eslint-plugin-import"],
```
Comment (mangs) [inline lib/eslintTypescriptConfig.json:6]: `eslint-plugin-import` is added in the base config

File: lib/eslintTypescriptConfig.json [hunk 3]
```diff
@@ -3,14 +3,13 @@
...
+  "plugins": ["@typescript-eslint/eslint-plugin"],
   "extends": [
-    "plugin:import/typescript",
     "eslint-config-airbnb-typescript/base",
     "plugin:@typescript-eslint/recommended-type-checked",
     "plugin:@typescript-eslint/stylistic-type-checked",
     "./eslintBaseConfig.json",
-    "eslint-config-prettier"
+    "plugin:import/typescript"
```
Comment (mangs) [inline lib/eslintTypescriptConfig.json:12]: adding to `eslint-plugin-import` should only happen after it is first added to `extends` (in `eslintBaseConfig.json`)

File: package.json [hunk 4]
```diff
@@ -1,21 +1,24 @@
...
+  "version": "2.0.0",
   "author": "Eric L. Goldstein <egoldstein@babbel.com>",
   "description": "Hierarchical ESLint configuration collection that intends to be simple to use, layered, and shared with others",
   "keywords": [
     "eslint",
     "eslintconfig"
   ],
   "engines": {
-    "node": ">=16.13.0"
+    "node": "^18.18.0 || >=20.0.0"
```
Comment (mangs) [inline package.json:11]: semver breaking change

File: package.json [hunk 5]
```diff
@@ -28,53 +31,60 @@
...
-    "react": "18.2.0",
-    "react-dom": "18.2.0",
-    "typescript": "5.3.3"
+    "react": "18.3.1",
+    "react-dom": "18.3.1",
+    "typescript": "5.5.3"
   },
   "peerDependencies": {
-    "eslint": ">=8.44.0"
+    "eslint": "^8.56.0"
```
Comment (mangs) [inline package.json:87]: semver breaking change (version bump, and eslint 9 is not allowed due to airbnb incompatibility)

File: test/eslintPlaywrightConfig/.eslintrc.json [hunk 6]
```diff
@@ -1,4 +1,4 @@
 {
   "root": true,
-  "extends": ["../../lib/eslintPlaywrightConfig.json"]
+  "extends": ["../../lib/eslintPlaywrightConfig.json", "../../lib/eslintNodeConfig.json"]
```
Comment (mangs) [inline test/eslintPlaywrightConfig/.eslintrc.json:3]: required for test to pass

File: tsconfig.json [hunk 7]
```diff
@@ -1,45 +1,42 @@
...
   "$schema": "https://json.schemastore.org/tsconfig",
   "compilerOptions": {
-    "allowImportingTsExtensions": true,
-    "verbatimModuleSyntax": true,
-    "jsx": "preserve",
+    "allowJs": false,
+    "module": "Preserve",
 
-    // ASTRO BASE TSCONFIG
+    // ASTRO BASE TSCONFIG (https://github.com/withastro/astro/blob/main/packages/astro/tsconfigs/base.json)
```
Comment (mangs) [inline tsconfig.json:7]: here down is a copy of all Astro's tsconfigs

File: tsconfig.json [hunk 8]
```diff
@@ -1,45 +1,42 @@
 {
   "$schema": "https://json.schemastore.org/tsconfig",
   "compilerOptions": {
-    "allowImportingTsExtensions": true,
-    "verbatimModuleSyntax": true,
-    "jsx": "preserve",
+    "allowJs": false,
+    "module": "Preserve",
```
Comment (mangs) [inline tsconfig.json:5]: Bun-compatible module mode (`import` and `require` in the same file)

**Review by mangs — COMMENTED** [review]
Some notes for context:

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,5 +1,17 @@
 # Changelog
 
+## 2.0.0
```
Comment (jduthon) [inline CHANGELOG.md:3]: Maybe we could highlight the breaking changes here for ease of update?

File: lib/eslintBaseConfig.json [hunk 10]
```diff
@@ -39,10 +38,9 @@
       {
         "groups": [
           ["builtin", "external"],
-          ["index", "internal", "parent", "sibling"],
+          ["index", "internal", "parent", "sibling", "unknown"],
```
Comment (jduthon) [inline lib/eslintBaseConfig.json:41]: Changes in this file don't seem to be documented in the Changelog

File: package.json [hunk 11]
```diff
@@ -28,53 +31,60 @@
...
+    "install:bun:expected-version": "bun --print '(await import(Bun.env.npm_package_json)).packageManager.split(\"@\")[1]' | (read BUN_VERSION; if [ $BUN_VERSION == `bun --version` ]; then echo Bun version $BUN_VERSION is already installed; else echo Installing Bun version $BUN_VERSION... && curl -fsSL https://bun.sh/install | bash -s \"bun-v$BUN_VERSION\"; fi)",
     "list:eslint:disable-directives": "rg '/(/|\\*+)[ \\t]*eslint-disable[^*]*(\\*+/)?'",
-    "list:todo-comments": "rg --only-matching '(TODO|FIXME):[a-zA-Z0-9\\t .,;?]+'",
-    "reinstall": "npm run --silent delete:node_modules && npm run --silent delete:package-lock && npm i",
-    "reinstall:use-lock-file": "npm run --silent delete:node_modules && npm ci",
-    "test": "eslint --max-warnings 0 --report-unused-disable-directives test/**/*.{js,jsx,mjs,ts,tsx}",
-    "validate:formatting": "prettier --check --no-editorconfig .",
-    "validate:linting:eslint": "eslint-config-prettier ./lib/eslintBaseConfig.json"
+    "list:todo-comments": "rg --only-matching '(TODO|FIXME)([\t ]+\\[[^\\]]+\\])?:[\\w\\s!\"#$%&'\\''()+,./:;<=>?@\\^_`{|}~-]+'",
+    "postinstall": "bun run --silent check:environment:versions",
```
Comment (jduthon) [inline package.json]: I think you didn't intend this, but `postinstall` scripts are also ran on consumers' machine, so with those changes, if one tries to run `npm i @babbel/eslint-config`, after installing the package, npm will attempt to run this script.

Screenshot on my machine after publishing to Verdaccio:
Screenshot 2024-07-24 at 16 05 24

**Review by jduthon — COMMENTED** [review]
Thanks a lot for the added configs and the package updates 👏 

I added two minor comments, I think the only one we really need to look into is the `postinstall` script as I think its implications are unintended, otherwise ready to go!

File: package.json (hunk 11 above)
Comment (mangs) [reply package.json]: Oh no, I forgot about that. Thank you for pointing that out! I was intending to ensure the proper Bun version was used for maintaining this library.

Comment (mangs) [reply package.json]: I tried a few ways to keep it in, but I ended up removing it

File: CHANGELOG.md (hunk 9 above)
Comment (mangs) [reply CHANGELOG.md:3]: Sure thing, adding

File: lib/eslintBaseConfig.json (hunk 10 above)
Comment (mangs) [reply lib/eslintBaseConfig.json:41]: Added

File: package.json (hunk 11 above)
Comment (jduthon) [reply package.json]: Could you have forgotten to push? I don't seem to see any changes

Comment (mangs) [reply package.json]: I pushed to `origin` (my fork) not `upstream`. I just pushed now. Thanks for pointing that out.

**Review by jduthon — APPROVED** [review]
Thanks for the changes!
Looking great 🚀

**Pull request merged**